File: order_service/src/main/java/com/dineflow/order/service/ReservationService.java
package com.dineflow.order.service;

import com.dineflow.order.dto.ReservationRequest;
import com.dineflow.order.dto.ReservationResponse;
import com.dineflow.order.exception.ResourceNotFoundException;
import com.dineflow.order.exception.TableAlreadyReservedException;
import com.dineflow.order.model.Reservation;
import com.dineflow.order.model.RestaurantTable;
import com.dineflow.order.model.enums.ReservationStatus;
import com.dineflow.order.repository.ReservationRepository;
import com.dineflow.order.repository.RestaurantTableRepository;

import lombok.RequiredArgsConstructor;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class ReservationService {

        private final ReservationRepository reservationRepository;
        private final RestaurantTableRepository tableRepository;

        @Transactional
        public ReservationResponse createReservation(ReservationRequest request) {

                // Validate time range
                if (!request.endTime().isAfter(request.startTime())) {
                        throw new IllegalArgumentException("End time must be after start time");
                }

                // Find table
                RestaurantTable table = tableRepository.findById(request.tableId())
                                .orElseThrow(() -> new ResourceNotFoundException(
                                                "Table not found: " + request.tableId()));

                // Check for overlapping reservations
                boolean alreadyBooked = reservationRepository.existsOverlappingReservation(
                                table.getId(),
                                request.date(),
                                request.startTime(),
                                request.endTime(),
                                List.of(
                                                ReservationStatus.PENDING,
                                                ReservationStatus.CONFIRMED));

                if (alreadyBooked) {
                        throw new TableAlreadyReservedException(
                                        "Table is already reserved for the selected date and time");
                }

                // Create reservation
                Reservation reservation = Reservation.builder()
                                .reservationReference(generateReservationReference())
                                .date(request.date())
                                .startTime(request.startTime())
                                .endTime(request.endTime())
                                .customerName(request.customerName())
                                .phoneNumber(request.phoneNumber())
                                .table(table)
                                .status(ReservationStatus.PENDING)
                                .createdAt(LocalDateTime.now())
                                .build();

                Reservation saved = reservationRepository.save(reservation);

                return toResponse(saved);
        }

        @Transactional(readOnly = true)
        public ReservationResponse getReservation(Long id) {

                Reservation reservation = reservationRepository.findById(id)
                                .orElseThrow(() -> new ResourceNotFoundException(
                                                "Reservation not found: " + id));

                return toResponse(reservation);
        }

        @Transactional(readOnly = true)
        public List<ReservationResponse> getReservationsByDate(LocalDate date) {

                return reservationRepository
                                .findByDateOrderByStartTimeAsc(date)
                                .stream()
                                .map(this::toResponse)
                                .toList();
        }

        @Transactional(readOnly = true)
        public ReservationResponse getReservationByReference(String reference) {

                Reservation reservation = reservationRepository
                                .findByReservationReference(reference)
                                .orElseThrow(() -> new ResourceNotFoundException(
                                                "Reservation not found: " + reference));

                return toResponse(reservation);
        }

        @Transactional(readOnly = true)
        public List<ReservationResponse> getAllReservations() {

                return reservationRepository
                                .findAll()
                                .stream()
                                .map(this::toResponse)
                                .toList();
        }

        @Transactional
        public ReservationResponse updateStatus(
                        Long id,
                        ReservationStatus status) {

                Reservation reservation = reservationRepository.findById(id)
                                .orElseThrow(() -> new ResourceNotFoundException("Reservation not found: " + id));

                reservation.setStatus(status);
                Reservation updated = reservationRepository.save(reservation);

                return toResponse(updated);
        }

        private ReservationResponse toResponse(Reservation reservation) {

                return new ReservationResponse(
                                reservation.getId(),
                                reservation.getReservationReference(),
                                reservation.getDate(),
                                reservation.getStartTime(),
                                reservation.getEndTime(),
                                reservation.getTable().getId(),
                                reservation.getTable().getTableNumber(),
                                reservation.getCustomerName(),
                                reservation.getPhoneNumber(),
                                reservation.getStatus(),
                                reservation.getCreatedAt());
        }

        private String generateReservationReference() {

                return "RES-" +
                                UUID.randomUUID()
                                                .toString()
                                                .substring(0, 8)
                                                .toUpperCase();
        }
}

File: frontend/src/service/reservationService.ts
import { orderClient } from "./api";

import type {
  RestaurantTable,
  ReservationRequest,
  ReservationResponse,
  ReservationStatus,
} from "../types/reservation";

export const getTables = async (): Promise<RestaurantTable[]> => {
  const response = await orderClient.get<RestaurantTable[]>(
    "/api/tables"
  );

  return response.data;
};

export const getTable = async (
  id: number
): Promise<RestaurantTable> => {
  const response = await orderClient.get<RestaurantTable>(
    `/api/tables/${id}`
  );

  return response.data;
};

export const createReservation = async (
  request: ReservationRequest
): Promise<ReservationResponse> => {
  const response = await orderClient.post<ReservationResponse>(
    "/api/reservations",
    request
  );

  return response.data;
};

export const getReservationById = async (
  id: number
): Promise<ReservationResponse> => {
  const response = await orderClient.get<ReservationResponse>(
    `/api/reservations/${id}`
  );

  return response.data;
};

export const getReservationsByDate = async (
  date: string
): Promise<ReservationResponse[]> => {
  const response = await orderClient.get<ReservationResponse[]>(
    "/api/reservations",
    {
      params: { date },
    }
  );

  return response.data;
};

export const getReservationByReference = async (
    reference: string
  ): Promise<ReservationResponse> => {
    const response = await orderClient.get<ReservationResponse>(
        `/api/reservations/reference/${reference}`
    );
  
    return response.data;
  };

export const getAllReservations = async (): Promise<
  ReservationResponse[]
> => {
  const response = await orderClient.get<ReservationResponse[]>(
    "/api/reservations"
  );

  return response.data;
};

export const updateReservationStatus = async (
  id: number,
  status: ReservationStatus
): Promise<ReservationResponse> => {
  const response = await orderClient.patch<ReservationResponse>(
    `/api/reservations/${id}/status`,
    null,
    {
      params: { status },
    }
  );

  return response.data;
};